### PR TITLE
Breaking: Strict package exports

### PR DESCRIPTION
**Pull request by nzakas**
## Summary

Node.js 12 introduced the ability to strictly define [package exports](https://nodejs.org/api/packages.html#packages_exports) that ensure internal package modules cannot be accessed externally. This RFC proposes implementing package exports for the `eslint` package to prevent unexpected use of internal modules.

## Related Issues

https://github.com/eslint/eslint/issues/13654

### CODE REVIEW

File: designs/2021-package-exports/README.md
Comment (ljharb) [reply designs/2021-package-exports/README.md:106]: I agree with every one of these points.

Comment (aladdin-add) [reply designs/2021-package-exports/README.md:106]: maybe we can go further: move the core rules to another package (e.g. @eslint/core-rules), just like the `eslint --init`([rfcs#79](https://github.com/eslint/rfcs/pull/79))

File: designs/2021-package-exports/README.md
Comment (nzakas) [reply designs/2021-package-exports/README.md:115]: Awesome, thanks for the insights.

File: designs/2021-package-exports/README.md
Comment (nzakas) [reply designs/2021-package-exports/README.md:106]: > Rules are already considered part of ESLint's public API... 

This is just untrue. People have been using rules as if they were part of the public API for a while, and we've always discouraged people from doing this but we had no way to enforce it until now. We simply can't support rules the same way we support documented APIs. 

> Rules are already considered part of ESLint's public API (ofc, as users consume them via configs), and rule removals are already a breaking change as per eslint's current semver policy. So it shouldn't be an additional maintenance burden to expose rules in this manner.

This is an assumption you're making and not a fact. Yes, removing a rule is a breaking change, but that's not the only type of change we can make. We can require additional meta data or remove existing meta data, we can change the function signature, we can change the overall format (as we already did, from function to object), we can change the messages being produced, we can change the directory structure. The rules are an unstable interface and only designed to be used inside of ESLint.

Each rule we maintain has a maintenance burden, and if we add compatibility with projects who are using rules in ways we don't expect, then that is more of a burden than otherwise.

> The issue with completely forking the rule is that it is a very heavy, labor intensive process - it would require us to explicitly monitor the upstream codebase and manually merge each and every change made to the rules.

I'm not sure I understand. Wouldn't you just wait until a new ESLint release rather than monitoring every commit?

Also, I'm not clear on how using a copy of the rule, rather than referencing it from ESLint, could lead to incompatibilities. Take for example [`comma-dangle`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/comma-dangle.ts). This file [references the original ESLint rule](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/comma-dangle.ts#L2) and then uses that to create the TypeScript-specific version. It seems like you could accomplish this with copies of rules, but just copying them into a specific directory and referencing them the same way (so `require("./original-rules/comma-dangle")` instead of `require("eslint/lib/rules/comma-dangel")`).

I understand that perhaps there would be a lag, though again, it seems like it would be a fairly simply copy operation since you are already referencing (without modification) the rules ESLint is shipping with.

Comment (JamesHenry) [reply designs/2021-package-exports/README.md:106]: @nzakas What you are suggesting would couple typescript-eslint to an exact patch version of eslint, which is a massive change from now where the user is in control of the ESLint that is used at runtime

**Comment by nzakas** [comment]
Thanks for everyone's feedback, after thinking about things some more, and discussing with the TSC, we've decided to make some changes to this proposal.

First and foremost we need to be clear: the TSC does not believe that supporting core rules as part of our public API is a good or desired outcome for all of the reasons I've already mentioned. However, it seems pretty clear that those who are already accessing core rules through undocumented APIs are willing to accept that this is a "void your warranty" situation and they are responsible for monitoring for breaking changes and responding to them without involving the ESLint team.

As a result, this latest version of the RFC introduces a new entrypoint called `use-at-your-own-risk`, where we will publish APIs that are not supported or guaranteed in any way. As part of that, there will be a `rules` API that grants access to ESLint's core rules. Because `rules` is a `Map`, it removes the dependence on the file system and directory structure to allow access to rules, which means that consumers will no longer have to worry about what happens if we change the directory structure. If we find other undocumented APIs that are critical to remain available, we can also add those `use-at-your-own-risk` with the same caveats.

Again, to be clear, anything in the `use-at-your-own-risk` entrypoint is considered unsupported and may contain breaking changes at any point in time. It's up to consumers of these APIs to monitor ESLint releases for possible breakages.

**Review by btmills — APPROVED** [review]
If the package maintainers who consume the ESLint API are happy with this version, then it works for me too.

**Comment by dbaeumer** [comment]
What is your schedule to release v8.0.0. The biggest problem I have with removing the old CLIEngine is that I as an VS Code extension owner (e.g. the ESLint extension for VS Code) still have to support the old API. Users can still rely on ESLint version < 8.0.0 in their workspace which will not have the new API especially not ESLint#getRulesForReport(report) So I need to write probing code and then do the right thing depending on the ESLint version installed. This will be quite some code effort on my end.

Can the old API still be accessible via the `se-at-your-own-risk` entry point?

**Comment by nzakas** [comment]
@dbaeumer can you check to see if CLIEngine is present and if not, use ESLint? Or is there some complexity here that I’m not seeing? (We are exporting a CommonJS module so I’m assuming that such a check is fairly easy, but please let me know if I’m missing something about your use case.)

**Comment by nzakas** [comment]
Oops, I realized I forgot to answer the first question. The timeline for v8.0.0 is not set, but my guess is most likely around two months from now. This RFC must remain open for at least two more weeks and we also need to finish up the other tasks, including an Espree update.

**Comment by dbaeumer** [comment]
@nzakas  Checking for CLIEngine versus ESLint is not complicated. But to my knowledge the new API is async whereas the old one is sync. So there is more than a simple if :-)

**Comment by dbaeumer** [comment]
And thanks for the timeline forecast.

**Comment by nzakas** [comment]
@dbaeumer can you explain more about why the change to async is problematic? I’m envisioning that you could normalize this by creating a CLIEngine wrapper that just wraps the method call return values with Promise.resolve(). I’d be happy to help with such a transition.

**Comment by dbaeumer** [comment]
It is not problematic. It is simply coding effort. And usually it ripples through the call stack (functions that call async API need to become async as well).

**Comment by nzakas** [comment]
@dbaeumer Gotcha. Well, as I mentioned, I'm happy to help with that transition to ease the burden on you. I'll be in touch when the first v8.0.0 prerelease is out and we can go from there.

**Comment by nzakas** [comment]
Moving go final commenting.

**Comment by ljharb** [comment]
Given that rules must operate synchronously, an api moving to async means that it can’t be used by rules.

**Comment by nzakas** [comment]
@ljharb That is a true system but i can’t imagine a scenario where a rule would use the `ESLint` class.

**Review by ljharb — COMMENTED** [review]
Please be sure `"./package.json": "./package.json"` is in "exports", otherwise every tool that looks at eslint's version (eslint-plugin-import, eslint-plugin-react, eslint-plugin-jsx-a11y, etc) will break, as well as most bundlers.

eslint-plugin-import currently uses these APIs in rules:
  - `require('eslint/lib/cli-engine/file-enumerator').FileEnumerator`
 
 eslint-plugin-react currently uses these APIs in tests, using RuleTester.linter.defineRule:
  - `eslint/lib/rules/no-undef`
  - `eslint/lib/rules/no-unused-vars`
  - `eslint/lib/rules/prefer-const`
  
  eslint-find-rules uses these APIs:
 - `eslint/lib/shared/relative-module-resolver`
 - `require('@eslint/eslintrc').Legacy.naming`

(The things here that use rules only need an opaque object that can be passed to defineRule)

**Comment by nzakas** [comment]
@ljharb thank you. 🙏 I've updated the RFC to reflect these. 

The `relativeModuleResolver` is now available in the `@eslint/eslintrc` package as `ModuleResolver`, so that should cover `eslint-find-rules`.

Is there a reason `eslint-plugin-import` uses `FileEnumerator` instead of a general-purpose globber? This will ultimately go away when we move to the new config as we will be able to use something like `globby` instead.

**Comment by ljharb** [comment]
@nzakas the reason is because we want to always use the exact identical globbing method eslint is using, which includes skipping eslintignore - there's no guarantee a general-purpose globber will match what eslint does.

Any solution for the `defineRule` RuleTester case?

**Comment by aladdin-add** [comment]
`RuleTester` is a public API, so it is usable. am I missing something?

**Comment by nzakas** [comment]
@ljharb hmmm okay, we should talk more about that down the line because I'd like to get rid of FileEnumerator and just use something off-the-shelf.

And as @aladdin-add mentioned, `RuleTester` is already part of the public API and remains so, and `eslint-plugin-react` can use the `use-at-your-own-risk` entrypoint to access the rules it's using.

**Comment by nzakas** [comment]
The final commenting period has ended, so merging. We can iterate more when there is a v8.0.0 prototype to try out and see what issues we run into.

**Comment by ljharb** [comment]
@aladdin-add RuleTester is fine, it’s that defineRule needs to take a rule object.

**Pull request merged**
